### PR TITLE
[api] Add ratelimit for websocket API

### PR DESCRIPTION
**Pull request by millken**
# Description
This PR introduces a rate limit for the websocket API to control the maximum number of messages a client can send per second. This is to ensure the stability and performance of the API by preventing any potential abuse or overload.

```
api:
  ...
  websocketMaxRateMessages: 5 #the maximum number of messages per second per client
```

Fixes #(issue)

## Type of change
Please delete options that are not relevant.
- [] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [] Code refactor or improvement
- [] Breaking change (fix or feature that would cause a new or changed behavior of existing functionality)
- [] This change requires a documentation update

# How Has This Been Tested?
Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration
- [x] make test
- [] fullsync
- [] Other test (please specify)

**Test Configuration**:
- Firmware version:
- Hardware:
- Toolchain:
- SDK:

# Checklist:
- [] My code follows the style guidelines of this project
- [] I have performed a self-review of my code
- [] I have commented my code, particularly in hard-to-understand areas
- [] I have made corresponding changes to the documentation
- [] My changes generate no new warnings
- [] I have added tests that prove my fix is effective or that my feature works
- [] New and existing unit tests pass locally with my changes
- [] Any dependent changes have been merged and published in downstream modules

### CODE REVIEW

**Review by envestcc — COMMENTED** [review]
Can rate limit be implemented in gateway?

**Comment by millken** [comment]
> Can rate limit be implemented in gateway?

No, because the websocket is a long connection, the gateway cannot ratelimit requests inside the connection.

File: api/serverV2.go [hunk 1]
```diff
@@ -65,7 +66,8 @@ func NewServerV2(
 
 	wrappedWeb3Handler := otelhttp.NewHandler(newHTTPHandler(web3Handler), "web3.jsonrpc")
 
-	wrappedWebsocketHandler := otelhttp.NewHandler(NewWebsocketHandler(web3Handler), "web3.websocket")
+	limiter := rate.NewLimiter(rate.Limit(cfg.WebsocketMaxRateMessages), 1)
```
Comment (envestcc) [inline api/serverV2.go:69]: Does it means allowing 5 req/s and only 1 req at the same time?

Comment (millken) [reply api/serverV2.go:69]: > Does it means allowing 5 req/s and only 1 req at the same time?


once a client is connected, the WebSocket server will limit the request rate to 5 req/s for this client. There is no limit to the number of client sides connected to the server( can limit in gateway)

**Comment by Liuhaai** [comment]
https://github.com/gorilla/websocket/issues/628#issuecomment-678822620

File: api/websocket.go [hunk 2]
```diff
@@ -2,12 +2,14 @@ package api
...
 import (
 	"context"
+	"math"
 	"net/http"
 	"sync"
 	"time"
 
 	"github.com/gorilla/websocket"
 	"go.uber.org/zap"
+	"golang.org/x/time/rate"
```
Comment (Liuhaai) [inline api/websocket.go:12]: https://github.com/uber-go/ratelimit

Comment (millken) [reply api/websocket.go:12]: `golang.org/x/time/rate` already meets our needs, i think no need to import a third-party library

File: api/serverV2.go (hunk 1 above)
Comment (envestcc) [reply api/serverV2.go:69]: so, what does the second param `1` means?

Comment (millken) [reply api/serverV2.go:69]: > so, what does the second param `1` means?

The  second parameter determines the maximum burst events that the limiter can tolerate. For example, if the value of `b` is 10, then even if 10 events suddenly occur in a short time. set to `1` means that the rate limit is based on the setting of the first parameter, and too many bursts are not allowed.

File: api/config.go [hunk 3]
```diff
@@ -23,16 +23,19 @@ type Config struct {
 	Tracer          tracer.Config     `yaml:"tracer"`
 	// BatchRequestLimit is the maximum number of requests in a batch.
 	BatchRequestLimit int `yaml:"batchRequestLimit"`
+	// WebsocketMaxRateMessages is the maximum number of messages per second per client.
+	WebsocketMaxRateMessages int `yaml:"websocketMaxRateMessages"`
```
Comment (CoderZhi) [inline api/config.go:27]: WebsocketRateLimit int `yaml:"websocketRateLimit"`

**Review by CoderZhi — APPROVED** [review]

**Review by dustinxie — APPROVED** [review]

**Pull request merged**